### PR TITLE
Add `THREDDSMergedSource` implementation

### PR DESCRIPTION
**Pull request by martindurant**
xref https://github.com/intake/intake-xarray/issues/29 @rabernat

This example takes a THREDDS URL and a path to descend down, and calls the combine function on all of the datasets found, e.g., 
```
import intake
import intake_thredds
s = intake_thredds.source.THREDDSMergedSource('http://dap.nci.org.au/thredds/catalog.xml', ['eMAST TERN', 'eMAST TERN - files', 'ASCAT', 'ASCAT_v1-0_soil-moisture_daily_0-05deg_2007-2011', '00000000', '*.nc'])
s.to_dask()
```
results in (and this takes a while)
```
<xarray.Dataset>
Dimensions:                                 (lat: 681, lon: 841, time: 1826)
Coordinates:
  * lat                                     (lat) float64 -10.0 -10.05 ... -44.0
  * lon                                     (lon) float64 112.0 112.0 ... 154.0
  * time                                    (time) datetime64[ns] 2007-01-01 ... 2011-12-31
Data variables:
    crs                                     (time) uint8 129 129 129 ... 129 129
    lwe_thickness_of_soil_moisture_content  (time, lat, lon) float32 0.0 ... 0.0
```
(a randomly-chosen .nc file has 31 timepoints - days of a month, I think)

### CODE REVIEW

File: intake_thredds/source.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                data = [ds.to_dask() for ds in _match(cat, path)]
+            self._ds = xr.combine_by_coords(data)
+
+
+def _match(cat, patterns):
+    out = []
+    for name in cat:
+        if fnmatch.fnmatch(name, patterns[0]):
+            if len(patterns) == 1:
+                out.append(cat[name](chunks={}))
```
Comment (andersy005) [inline intake_thredds/source.py:67]: Using `chunks={}` appears to speed the xarray's combine up:

```python
In [1]: import intake                                                                                              

In [2]: url = 'http://dap.nci.org.au/thredds/catalog.xml'                                                          

In [3]: paths = ['eMAST TERN', 'eMAST TERN - files', 'ASCAT', 'ASCAT_v1-0_soil-moisture_daily_0-05deg_2007-2011', '
   ...: 00000000', '*.nc']                                                                                         

In [4]: s = intake.open_thredds_merged(url, paths)                                                                 

In [5]: %%time 
   ...: ds = s.to_dask() 
   ...:  
   ...:                                                                                                            
Dataset(s): 100%|██████████████████████████████| 60/60 [02:36<00:00,  2.60s/it]
CPU times: user 875 ms, sys: 235 ms, total: 1.11 s
Wall time: 2min 41s
```


Screen Shot 2020-04-17 at 4 02 47 PM

**Comment by andersy005** [comment]
@martindurant, I added a few changes to this PR. When you get a moment, can you take a look and let me know what you think.

Currently the data loading is dispatched to `netcdf source`, should we use the [`opendap` source](https://github.com/intake/intake-xarray/blob/master/intake_xarray/opendap.py) instead?

File: intake_thredds/source.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                if '*' not in part and '?' not in part:
+                    cat = cat[part]()
+                else:
+                    break
+            path = self.path[i:]
+            if self.progressbar:
+                data = [ds.to_dask() for ds in tqdm(_match(cat, path), desc='Dataset(s)', ncols=79)]
+            else:
+                data = [ds.to_dask() for ds in _match(cat, path)]
+            self._ds = xr.combine_by_coords(data)
```
Comment (andersy005) [inline intake_thredds/source.py:59]: The `auto_combine()` function was deprecated in the most recent versions of xarray.

**Comment by martindurant** [comment]
> should we use the opendap source instead

because of the auth options? I don't know the typical use for thredds, so I leave it up to you. The metadata describes the endpoint as dap, if I remember.

**Comment by aaronspring** [comment]
I would be interested in this PR merged. Any progress here? @andersy005 @martindurant @larsbuntemeyer

**Comment by martindurant** [comment]
I honestly don't remember where this was up to.
@aaronspring , you may be in a fine place to have an opinion.
cc @scottyhq @larsbuntemeyer for some similar discussion in intake-xarray

**Comment by martindurant** [comment]
Alternatively, happy to merge if there are no more comments. @aaronspring , sound good?

**Comment by andersy005** [comment]
> I honestly don't remember where this was up to.
> Alternatively, happy to merge if there are no more comments. @aaronspring , sound good?

👍🏽 for merging this as is, and addressing issues + adding new features in separate PRs....

**Pull request merged**